backend-demo/src/screencast.rs

```diff
@@ -0,0 +1,117 @@
+
+#[derive(Default)]
+pub struct Screencast {
+    sessions: Arc<Mutex<Vec<String>>>,
+}
+
+#[async_trait]
+impl RequestImpl for Screencast {
+    async fn close(&self, token: HandleToken) {
+        tracing::debug!("IN Close(): {token}");
+    }
+}
+
+#[async_trait]
+impl ScreencastImpl for Screencast {
+    fn available_source_types(&self) -> BitFlags<SourceType> {
+        SourceType::Monitor | SourceType::Window
+    }
+
+    fn available_cursor_mode(&self) -> BitFlags<CursorMode> {
+        CursorMode::Hidden | CursorMode::Embedded | CursorMode::Metadata
+    }
+
+    async fn create_session(
+        &self,
+        handle_token: HandleToken,
+        session_handle_token: HandleToken,
+        _app_id: Option<AppID>,
+        _options: CreateSessionOptions,
+    ) -> Result<CreateSessionResponse> {
+        tracing::debug!("IN Screencast::create_session(): {handle_token}, {session_handle_token}");
+        let session = session_handle_token.to_string();
+        let mut sessions = self.sessions.lock().unwrap();
+        if sessions.contains(&session) {
+            let errormsg = format!("A session with handle `{session}` already exists");
+            tracing::error!("{}", errormsg.as_str());
+            return Err(PortalError::Exist(errormsg));
+        }
+        sessions.push(session.clone());
+        Ok(CreateSessionResponse::new(session_handle_token))
+    }
+
+    async fn select_sources(
+        &self,
+        session_handle_token: HandleToken,
+        _app_id: Option<AppID>,
+        _options: SelectSourcesOptions,
+    ) -> Result<SelectSourcesResponse> {
+        tracing::debug!("IN Screencast::select_sources(): {session_handle_token}");
+        let session = session_handle_token.to_string();
+        let sessions = self.sessions.lock().unwrap();
+        if !sessions.contains(&session) {
+            let errormsg = format!("Unknown session: `{session}`");
+            tracing::error!("{}", errormsg.as_str());
+            return Err(PortalError::NotFound(errormsg));
+        }
+        Ok(SelectSourcesResponse {})
+    }
+
+    async fn start_cast(
+        &self,
+        session_handle_token: HandleToken,
+        _app_id: Option<AppID>,
+        _window_identifier: Option<WindowIdentifierType>,
+        _options: StartCastOptions,
+    ) -> Result<Streams> {
+        tracing::debug!("IN Screencast::start_cast(): {session_handle_token}");
+        let session = session_handle_token.to_string();
+        let sessions = self.sessions.lock().unwrap();
+        if !sessions.contains(&session) {
+            let errormsg = format!("Unknown session: `{session}`");
+            tracing::error!("{}", errormsg.as_str());
+            return Err(PortalError::NotFound(errormsg));
+        }
+        let streams = vec![StreamBuilder::new(42)
+            .source_type(SourceType::Monitor)
+            .build()];
+        Ok(StreamsBuilder::new(streams).build())
+    }
+}
+
+#[async_trait]
+impl SessionMonitor for Screencast {
```

Comment:
What is the difference between SessionMonitor and a potentially named SessionImpl trait? Seems the same to me

Reply:
Indeed, no difference. The pattern of naming traits with an `Impl` suffix feels a bit weird to me, but consistency is worth it, so I'm renaming `SessionMonitor`to `SessionImpl`.

Reply:
The naming comes from the GTK Rust bindings, where we define a trait named `$TypeImpl` for `Impl`ementing a custom `$Type` and I followed the same here, it is not great but good enough.

Reply:
Got it, thanks for the reference.